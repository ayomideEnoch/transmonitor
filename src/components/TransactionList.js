import React from 'react';
import PropTypes from 'prop-types';
import {Table} from 'reactstrap';
import {ethWeb3} from '../web3connection';

export function TransactionList(props) {
  if (props.blocks.length === 0 || !props.selectedBlock) return null;
  let sortedBlocks = props.blocks.sort((a, b) => b.number - a.number);
  let idx = sortedBlocks.findIndex(block => block.number ===  props.selectedBlock);
  if (idx === -1) return null;
  let transactions = sortedBlocks[idx].valueTransactions;
  return (<div id="transactions">
    <h3>Block #{props.selectedBlock}</h3>
    <h5>{transactions.length} transactions with value</h5>
    <a href="#blocks"><span aria-hidden="true">⬆</span>Back to Blocks list</a>
    <Table bordered striped hover responsive size="sm">
      <caption>List of transactions with value in a selected block</caption>
      <thead>
        <tr>
          <th>hash #</th>
          <th>from</th>
          <th>to</th>
          <th>value</th>
        </tr>
      </thead>
      <tbody>
      {transactions.map((transaction) => <Transaction {...transaction} key={transaction.hash} />)}
      </tbody>
    </Table>
    <a href="#blocks"><span aria-hidden="true">⬆</span>Back to Blocks list</a>
  </div>);
}

function format(text) {
  return `${text.substring(0, 7)}...`;
}

function Transaction(transaction) {
  return (<tr>
    <th><a href={`https://etherscan.io/tx/${transaction.hash}`} target="_blank">{format(transaction.hash)}</a></th>
    <th>{format(transaction.from)}</th>
    <th>{format(transaction.to)}</th>
    <th>{ethWeb3.fromWei(transaction.value, 'ether').toString()} Ether</th>
  </tr>);
}

TransactionList.propTypes = {
  blocks: PropTypes.array,
  selectedBlock: PropTypes.number,
};

export default TransactionList;
