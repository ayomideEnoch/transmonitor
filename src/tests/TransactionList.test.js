import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {BigNumber} from 'bignumber.js';
import TransactionList from '../components/TransactionList';

describe('TransactionList', function() {
  let mockPropsNoSelectedBlock = {
    blocks: [{}],
    selectedBlock: null,
  };

  let mockPropsNoBlock = {
    blocks: [],
    selectedBlock: null,
  };

  // the order should be 123458: 3/6, 123457: 1/5, 123456: 2/4
  let mockProps = {
    blocks: [
      {
        number: 123456,
        valueTransactions: [{
          hash: '0xa51d242d58030d110ad8579ba04174e861d4f87e432d2498bae9423a4c6d5ed8',
          from: '0x471138d67093de52d9d11084aad707ba0800a5c2',
          to: '0x2a0c0dbecc7e4d658f48e01e3fa353f44050c208',
          value: new BigNumber('16500000000000000000'),
        }, {
          hash: '0x8ce9659381e74d654d648f24eff5209a535987523925204a85ce29ad2e21755a',
          from: '0x3dda61b077660745a6a8bef0d112f6f32ac67bca',
          to: '0x5a7456983c188415cda917f117c981ee5207d84f',
          value: new BigNumber('33000000000'),
        }],
        totalTransactionsLength: 4,
        timestamp: 1519748998
      },
      {
        number: 123457,
        valueTransactions: [{
          hash: '0xdd00f5aaae1af883e40098c927a6a8dc230b941654c3a2a5d9ad4c79f79c558a',
          from: '0x68b1a89523b7ed11f499f36ba266c688401cdbc1',
          to: '0x8ad76a1bef25cd903b8c1de72af6d2eb30218c95',
          value: new BigNumber('199889030000000000'),
        }],
        totalTransactionsLength: 5,
        timestamp: 1519749008
      },
      {
        number: 123458,
        valueTransactions: [{
          hash: '0x8234d8a159441b09a08b6ff019085eef938d07bcca29bae907e8bbcaf762f19',
          from: '0x776d8bf339a5735e8c7a3a7bbdd2ceb5293949bf',
          to: '0xe7e9643905ca3a07144c6da64b3ad0671b4b1a93',
          value: new BigNumber('2023320110000000000'),
        }, {
          hash: '0xaa8274f0168a1a6712ef0927e590cc7305bf1e12e63abbd2d615ced9a67792d8',
          from: '0x7e2541f720e44b3fef6b81b2adb148471ee8b63d',
          to: '0x2a0c0dbecc7e4d658f48e01e3fa353f44050c208',
          value: new BigNumber('612062408121995274'),
        }, {
          hash: '0x216ea767f72df6d56cd14dcee0d3ff96d96beb9e30c4b53709c9c95d0f90edc1',
          from: '0x850e9d041de71069cfb38bab2716cba34325979b',
          to: '0x23be96a5cdefd4fcf8e153691e15d7b21f5b2ff4',
          value: new BigNumber('1988327790000000000'),
        }],
        totalTransactionsLength: 6,
        timestamp: 1519749108
      },
    ],
    selectedBlock: 123457,
  };

  it('renders without crashing when no selectedBlock', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TransactionList {...mockPropsNoSelectedBlock} />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('renders without crashing when no block', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TransactionList {...mockPropsNoBlock} />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TransactionList {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('render null when no selectedBlock', () => {
    const component = renderer.create(<TransactionList {...mockPropsNoBlock} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('render null when no blocks', () => {
    const component = renderer.create(<TransactionList {...mockPropsNoBlock} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('render correctly', () => {
    const component = renderer.create(<TransactionList {...mockProps} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
});
