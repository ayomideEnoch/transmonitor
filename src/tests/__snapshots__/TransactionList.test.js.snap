// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionList render correctly 1`] = `
<div
  id="transactions"
>
  <h3>
    Block #
    123457
  </h3>
  <h5>
    1
     transactions with value
  </h5>
  <a
    href="#blocks"
    onClick={[Function]}
  >
    <span
      aria-hidden="true"
    >
      ⬆
    </span>
    Back to Blocks list
  </a>
  <div
    className="table-responsive"
  >
    <table
      className="table table-sm table-bordered table-striped table-hover"
    >
      <caption>
        List of transactions with value in a selected block
      </caption>
      <thead>
        <tr>
          <th>
            hash #
          </th>
          <th>
            from
          </th>
          <th>
            to
          </th>
          <th>
            value
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th>
            <a
              href="https://etherscan.io/tx/0xdd00f5aaae1af883e40098c927a6a8dc230b941654c3a2a5d9ad4c79f79c558a"
              target="_blank"
            >
              0xdd00f...
            </a>
          </th>
          <th>
            0x68b1a...
          </th>
          <th>
            0x8ad76...
          </th>
          <th>
            0.19988903
             Ether
          </th>
        </tr>
      </tbody>
    </table>
  </div>
  <a
    href="#blocks"
    onClick={[Function]}
  >
    <span
      aria-hidden="true"
    >
      ⬆
    </span>
    Back to Blocks list
  </a>
</div>
`;

exports[`TransactionList render correctly when some block has 0 transaction with value 1`] = `
<div
  id="transactions"
>
  <h3>
    Block #
    123458
  </h3>
  <h5>
    0
     transactions with value
  </h5>
  <a
    href="#blocks"
    onClick={[Function]}
  >
    <span
      aria-hidden="true"
    >
      ⬆
    </span>
    Back to Blocks list
  </a>
  <div
    className="table-responsive"
  >
    <table
      className="table table-sm table-bordered table-striped table-hover"
    >
      <caption>
        List of transactions with value in a selected block
      </caption>
      <thead>
        <tr>
          <th>
            hash #
          </th>
          <th>
            from
          </th>
          <th>
            to
          </th>
          <th>
            value
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
  <a
    href="#blocks"
    onClick={[Function]}
  >
    <span
      aria-hidden="true"
    >
      ⬆
    </span>
    Back to Blocks list
  </a>
</div>
`;

exports[`TransactionList render null when no blocks 1`] = `null`;

exports[`TransactionList render null when no selectedBlock 1`] = `null`;
